src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -0,0 +1,448 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions
+{
+    using static EmbeddedSyntaxHelpers;
+    using static RegexHelpers;
+
+    using RegexToken = EmbeddedSyntaxToken<RegexKind>;
+    using RegexTrivia = EmbeddedSyntaxTrivia<RegexKind>;
+
+    /// <summary>
+    /// Produces tokens from the sequence of <see cref="VirtualChar"/> characters.  Unlike the native C# and VB
+    /// lexer, this lexer is much more tightly controlled by the parser.  For example, while C# can have trivia
+    /// on virtual every token, the same is not true for RegexTokens.  As such, instead of automatically lexing
+    /// out tokens to make them available for the parser, the parser asks for each token as necessary passing
+    /// the right information to indicate which types and shapes of tokens are allowed.
+    /// 
+    /// The tight coupling means that the parser is allowed direct control of the position of the lexer.
+    /// 
+    /// Note: most of the time, tokens returned are just a single character long, including for long sequences
```

Comment:
What would be an example of a multi-character token?

Reply:
I will add a comment.  but something like  ```a{10,20}```.  ```10``` and ```20``` will be multi character tokens.  ```a```, ```{```, ```,```, and ```}``` are all single character tokens.

Reply:
#resolved